crates/notifier_host/src/icon.rs

```diff
@@ -0,0 +1,196 @@
+use crate::*;
+
+use gtk::{self, prelude::*};
+
+#[derive(thiserror::Error, Debug)]
+pub enum IconError {
+    #[error("while fetching icon name: {0}")]
+    DBusIconName(#[source] zbus::Error),
+    #[error("while fetching icon theme path: {0}")]
+    DBusTheme(#[source] zbus::Error),
+    #[error("while fetching pixmap: {0}")]
+    DBusPixmap(#[source] zbus::Error),
+    #[error("loading icon from file {path:?}")]
+    LoadIconFromFile {
+        path: String,
+        #[source]
+        source: gtk::glib::Error,
+    },
+    #[error("loading icon {icon_name:?} from theme {}", .theme_path.as_ref().unwrap_or(&"(default)".to_owned()))]
+    LoadIconFromTheme {
+        icon_name: String,
+        theme_path: Option<String>,
+        #[source]
+        source: gtk::glib::Error,
+    },
+    #[error("no icon available")]
+    NotAvailable,
+}
+
+/// Get the fallback GTK icon
+pub async fn fallback_icon(size: i32) -> gtk::gdk_pixbuf::Pixbuf {
+    let theme = gtk::IconTheme::default().expect("Could not get default gtk theme");
+    return match theme.load_icon("image-missing", size, gtk::IconLookupFlags::FORCE_SIZE) {
+        Err(e) => {
+            log::error!("failed to load \"image-missing\" from default theme: {}", e);
+            // create a blank pixbuf
+            gtk::gdk_pixbuf::Pixbuf::new(gtk::gdk_pixbuf::Colorspace::Rgb, false, 0, size, size).unwrap()
+        }
+        Ok(pb) => pb.unwrap(),
+    };
+}
+
+/// Load a pixbuf from StatusNotifierItem's [Icon format].
+///
+/// [Icon format]: https://freedesktop.org/wiki/Specifications/StatusNotifierItem/Icons/
+fn icon_from_pixmap(width: i32, height: i32, mut data: Vec<u8>) -> gtk::gdk_pixbuf::Pixbuf {
+    // We need to convert data from ARGB32 to RGBA32, since that's the only one that gdk-pixbuf
+    // understands.
+    for chunk in data.chunks_mut(4) {
+        let a = chunk[0];
+        let r = chunk[1];
+        let g = chunk[2];
+        let b = chunk[3];
```

Comment:
I thiiiiink you can destructure here?

Reply:
unfortunately not really, since chunk is a `&mut [u8]` without a type-level guarantee of size

Reply:
I believe there's a try_into() implementation for conversion from a `&mut [u8]` to a `&mut [u8; 4]`. As a bonus, using try_into instead of indexing means there's only one error to handle rather than four separate ones (as each index operation can potentially panic).